### ISSUE TITLE
Math::BigInt::new argument format is too restrictive

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4915]
Migrated from [rt.perl.org#8348](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=8348) (status was 'resolved')

Searchable as RT8348$

**Comment by p5pRT** [comment]
### From @mjdominus

#### Created by @mjdominus



&nbsp;        perl \-MMath&#8203;::BigInt \-le '$z = 6004799503160661; $k = Math&#8203;::BigInt\->new\($z\); print $k'

&nbsp;        NaN

The problem here is that 6004799503160661 is too big to fit into an
integral type variable\.  But it's unfortunate that Math&#8203;::BigInt&#8203;::new
can't deal with this properly\.

I can get around the problem using trickery&#8203;:

&nbsp;        perl \-MMath&#8203;::BigInt \-le '$z = 6004799503160661; $k = Math&#8203;::BigInt\->new\(sprintf "%\.0f"\, $z\); print $k'

&nbsp;        \+6004799503160661

This is a shame\.  I wonder if Math&#8203;::BigInt&#8203;::new could be extended to
deal correctly with this case\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium

Site configuration information for perl v5.6.1:

Configured by root at Sat Dec 29 11:54:59 EST 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.4.2-2, archname=i586-linux
    uname='linux plover.com 2.4.2-2 #1 sun apr 8 19:37:14 edt 2001 i586 unknown '
    config_args='-des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing'
    ccversion='', gccversion='2.96 20000731 (Red Hat Linux 7.1 2.96-81)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.2.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /usr/local/lib/perl5/5.6.1/i586-linux
    /usr/local/lib/perl5/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.1/i586-linux
    /usr/local/lib/perl5/site_perl/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.0/i586-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home/mjd
    LANG=en_US
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/lib:/usr/lib:/usr/X11R6/lib
    LOGDIR (unset)
    PATH=/home/mjd/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/games:/sbin:/usr/sbin:/usr/local/bin/X11R6:/usr/local/bin/mh:/data/mysql/bin:/usr/local/bin/pbm:/usr/local/bin/ezmlm:/home/mjd/TPI/bin:/usr/local/teTeX/bin:/usr/local/mysql/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @JohnPeacock

mjd@&#8203;plover\.com wrote&#8203;:

> This is a bug report for perl from mjd@&#8203;plover\.com\,
> generated with the help of perlbug 1\.33 running under perl v5\.6\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
>         perl \-MMath&#8203;::BigInt \-le '$z = 6004799503160661; $k = Math&#8203;::BigInt\->new\($z\); print $k'
> 
>         NaN
> 
> The problem here is that 6004799503160661 is too big to fit into an
> integral type variable\.  But it's unfortunate that Math&#8203;::BigInt&#8203;::new
> can't deal with this properly\.


perl \-MMath&#8203;::BigInt \-le 'print $Math&#8203;::BigInt&#8203;::VERSION'
1\.49

perl \-MMath&#8203;::BigInt \-le '$z = 6004799503160661; $k = 
Math&#8203;::BigInt\->new\($z\); print $k'
6004799503160660

Tels already fixed this \(mostly ;~\)\.  The problem with your solution is 
that you already know $z is stored as a NV\, so washing the entry through 
&nbsp;  sprintf\(\) works for you\.  But\, as you can see\, on my system anyway\, 
the last decimal place is not maintained \(because of floating point 
notation error\)\.  This may be something Tels can fix better\, but no one 
can ever defeat the limiting nature of floating point notation\.

However\, you will see that

perl \-MMath&#8203;::BigInt \-le '$z = "6004799503160661"; $k = 
Math&#8203;::BigInt\->new\($z\); print $k'
6004799503160661

works as designed\.  Extremely large numbers \(i\.e\. over UV\_MAX\) are best 
passed as strings to new\, at which point they can be treated as numbers\. 
&nbsp;  Until we get automatic upgrade of IV=>UV=>Math&#8203;::BigInt at the tokizer 
level\, this is probably a "don't do that" rather than a bug\.

John

\-\- 
John Peacock
Director of Information Research and Technology
Rowman & Littlefield Publishing Group
4720 Boston Way
Lanham\, MD 20706
301\-459\-3366 x\.5010
fax 301\-429\-5747

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

\-\-\-\-\-BEGIN PGP SIGNED MESSAGE\-\-\-\-\-

Moin\,

>mjd@&#8203;plover\.com wrote&#8203;:
>
>> This is a bug report for perl from mjd@&#8203;plover\.com\,
>> generated with the help of perlbug 1\.33 running under perl v5\.6\.1\.
>>
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \[Please enter your report here\]
>>
>>         perl \-MMath&#8203;::BigInt \-le '$z = 6004799503160661; $k =
>Math&#8203;::BigInt\->new\($z\); print $k'
>>
>>         NaN
>>
>> The problem here is that 6004799503160661 is too big to fit into an
>> integral type variable\.  But it's unfortunate that Math&#8203;::BigInt&#8203;::new
>> can't deal with this properly\.
>
>
>perl \-MMath&#8203;::BigInt \-le 'print $Math&#8203;::BigInt&#8203;::VERSION'
>1\.49
>
>perl \-MMath&#8203;::BigInt \-le '$z = 6004799503160661; $k =
>Math&#8203;::BigInt\->new\($z\); print $k'
>6004799503160660
>
>Tels already fixed this \(mostly ;~\)\.  The problem with your solution is
>that you already know $z is stored as a NV\, so washing the entry through
>  sprintf\(\) works for you\.  But\, as you can see\, on my system anyway\,
>the last decimal place is not maintained \(because of floating point
>notation error\)\.  This may be something Tels can fix better\, but no one
>can ever defeat the limiting nature of floating point notation\.
>
>However\, you will see that
>
>perl \-MMath&#8203;::BigInt \-le '$z = "6004799503160661"; $k =
>Math&#8203;::BigInt\->new\($z\); print $k'
>6004799503160661
>
>works as designed\.  Extremely large numbers \(i\.e\. over UV\_MAX\) are best
>passed as strings to new\, at which point they can be treated as numbers\.
>  Until we get automatic upgrade of IV=>UV=>Math&#8203;::BigInt at the tokizer
>level\, this is probably a "don't do that" rather than a bug\.

John is right\, this is not a bug in MBI\, but more a bug in Perl\. When $z
reaches BigInt\->new\(\)\, it was already "converted" to an integer or float
constant by Perl\.

It is even documented in the POD of Math&#8203;::BigInt ;\) You can either use
strings \(by quoting\) or use Math&#8203;::BigInt '&#8203;:constant';

For fun\, just try&#8203;:

&nbsp;        use Math&#8203;::BigInt;

&nbsp;        print Math&#8203;::BigInt\->new\(2 \*\* 512\);

It won't work\, since 2 \*\* 512 is calculated as constant and the put into
BigInt\, where it is already too late to preserve all the precious bits\.

Hope this helps\,

Tels

PS&#8203;: Please note that modern versions of BigInt already take input like
"1\.23E\+2" and convert it correctly to 123\. This was an idea from John\, and
as long as the input is an valid integer \(e\.g\. no digits after the dot\)\, it
should work out ok\. 

\- \-\-
perl \-MMath&#8203;::String \-e 'print \\
Math&#8203;::String\->from\_number\("215960156869840440586892398248"\)\,"\\n"'

 http&#8203;://bloodgate\.com/perl       My current Perl projects
 PGP key available on http&#8203;://bloodgate\.com/tels\.asc or via email\.


\-\-\-\-\-BEGIN PGP SIGNATURE\-\-\-\-\-
Version&#8203;: 2\.6\.3i
Charset&#8203;: latin1

iQEVAwUBPFSUk3cLPEOTuEwVAQGx5Af\+KJUuimtATvjXBUqx3pL7DDNurvtDocCM
RDZdEthwWLTavF04I0FBdzK6GZgMv/o\+9zR0H9SpR5vg3SiVuMYz0SXejdM7idbS
8HJyAUmPz3FAweMyqdeE/auaQWOcth4PdXhx6v0JzRpoK5m/hV3of35YWgmyttUR
eo\+JOw8YcXhDEnQ/YZ1UsWndtxtAEsahTYcP2Z1fEUrzLOU6/O8OoRZ8iYA3PMvy
36w6aT9xSLgriVSjHjWmnZQZWGJnP29/0IOsU/ORSf2gY3on8qKJbjF8ME9WreaI
XQakEdtfZfs1PhwkYqX21u1c2zQonhASXlaVVjqdRGH19Zctgn6huQ==
=Epy0
\-\-\-\-\-END PGP SIGNATURE\-\-\-\-\-

**Comment by p5pRT** [comment]
### From @mjdominus

#### Created by @mjdominus



&nbsp;        perl \-MMath&#8203;::BigInt \-le '$z = 6004799503160661; $k = Math&#8203;::BigInt\->new\($z\); print $k'

&nbsp;        NaN

The problem here is that 6004799503160661 is too big to fit into an
integral type variable\.  But it's unfortunate that Math&#8203;::BigInt&#8203;::new
can't deal with this properly\.

I can get around the problem using trickery&#8203;:

&nbsp;        perl \-MMath&#8203;::BigInt \-le '$z = 6004799503160661; $k = Math&#8203;::BigInt\->new\(sprintf "%\.0f"\, $z\); print $k'

&nbsp;        \+6004799503160661

This is a shame\.  I wonder if Math&#8203;::BigInt&#8203;::new could be extended to
deal correctly with this case\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium

Site configuration information for perl v5.6.1:

Configured by root at Sat Dec 29 11:54:59 EST 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.4.2-2, archname=i586-linux
    uname='linux plover.com 2.4.2-2 #1 sun apr 8 19:37:14 edt 2001 i586 unknown '
    config_args='-des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing'
    ccversion='', gccversion='2.96 20000731 (Red Hat Linux 7.1 2.96-81)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.2.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /usr/local/lib/perl5/5.6.1/i586-linux
    /usr/local/lib/perl5/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.1/i586-linux
    /usr/local/lib/perl5/site_perl/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.0/i586-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home/mjd
    LANG=en_US
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/lib:/usr/lib:/usr/X11R6/lib
    LOGDIR (unset)
    PATH=/home/mjd/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/games:/sbin:/usr/sbin:/usr/local/bin/X11R6:/usr/local/bin/mh:/data/mysql/bin:/usr/local/bin/pbm:/usr/local/bin/ezmlm:/home/mjd/TPI/bin:/usr/local/teTeX/bin:/usr/local/mysql/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>
